### PR TITLE
iOS: Migrate FlutterEngine to ARC

### PR DESCRIPTION
**Pull request by cbracken**
Migrates `FlutterEnging` from manual reference counting to ARC. Migrates properties from `retain` to strong and `assign` to `weak` (where referencing an Obj-C object).

No semantic changes, therefore no changes to tests.

Issue: https://github.com/flutter/flutter/issues/137801

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] There weren't as many ARCane memory management gotchas in this patch as I'd have expected.
- [X] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:124]: And this one, in `setViewController` it can become `_viewController = viewController;`, and the `- (FlutterViewController*)viewController` getter implementation can be removed.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:310]: I _think_ these two are still needed?

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:124]: Yep! This is patch is still WIP -- it's at the "just barely not leaking anymore stage". There's a ton of cleanup still to come.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:310]: Oh crap -- yes ".parent". Thanks for spotting.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:157]: Added a TODO -- as discussed in person, this one is going to need to wait till `platform_view_ios.h` is migrated since we're going to need a `__weak` modifier on one of the C++ class members, and since that header is pulled into both FlutterEngine and FlutterViewController, we'd need to migrate FlutterViewController in this same patch.

Getting closer though!

**Comment by cbracken** [comment]
Apologies... this one is [a little bigger than the last few](https://www.youtube.com/watch?v=rFeVfwDvTyM).

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:105]: Thoughts? `copy`?

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:165]: These two have public getters for `NSObject<FlutterBinaryMessenger>` and `NSObject<FlutterTextureRegistryRelay>`. 

I'd prefer to hold off on these to a future patch and they're alright as-is for now.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:416]: This is just preserving the existing weak pointer semantics.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:506]: This is just preserving existing weak C++ lambda capture of `self`.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:718]: Here and a couple more below -- notice that I've moved these to use `weakSelf.platformPlugin`. 

Technically, this does very slightly change the semantics here, but for the better since if `self` were to be cleaned up, these will all devolve into no-ops.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm:85]: `_engine` here is a weak property.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.h
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.h:15]: We can make this (and probably some others) `(__weak FlutterEngine*)` once FlutterViewController.mm (which also #imports this) is migrated to ARC.

**Review by jmagman — APPROVED** [review]
LGTM once the `[labelPrefix copy]` is added to init.

This code gives me 30% less heebie jeebies than it did before.  Thank you so much for your service 🫡

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: Nice catch! Done.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:295]: TODO + tracking issue for eliminating this code by ensuring that `FlutterEngineRegistrar` is using `weak` pointers?

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:308]: Optional: It's harmless to leave, but this line hasn't been necessary since we dropped iOS 8 support. (The thing above looks like it probably is still because of the way it was registered.)

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (stuartmorgan) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:416]: Nit: it's much more common to call this variable `weakSelf`, and that's also more descriptive, so I would suggest using that convention.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:480]: Optional here and below: `url` is a property on `FlutterDartVMServicePublisher`, so this can be `self.publisher.url`.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:631]: Nit: technically it doesn't matter, but generally with the weakSelf/strongSelf pattern you only use `strongSelf` once it's created.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:796]: This code isn't safe if the comment you removed was wrong (and even if it happens to currently be safe, we shouldn't leave it this way) because the block is unconditionally using `strongSelf`. There should be a:

```objc
if (!strongSelf) {
  return;
}
```

right after this line.

**Review by stuartmorgan — APPROVED** [review]
🎉  LGTM with one or two fixes and some optional nits. (I did a full audit of all of the blocks in addition to reviewing the diff; I know you did as well, but just as a record that the area most likely to have subtle bugs did get more eyes.)

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:480]: Good call - I think my brain had turned to mush at this point. Done.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:631]: Oh good catch -- I wonder if I did the two blocks in different passes, but totally miss ed spotting this one. Done. Thanks!

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:796]: Nice catch -- did this in a couple others but totally missed it here. This block is expected to return a `std::unique_ptr<flutter::PlatformView>`, so returning a null one here in the hopes that whoever calls this is careful enough to check it (they are, at the moment):
https://github.com/flutter/engine/blob/e03449ed1ee1f18b06adb8c250dc0781be66db67/shell/common/shell.cc#L236-L239

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm]: Done.

File: shell/platform/darwin/ios/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:310]: Will followup in another PR just to "minimise" the change in this patch, which admittedly there is nothing minimal about at the moment. Mostly just want to work through why it's unnecessary, and it's currently to early in the morning for my brain to do that.

Comment (stuartmorgan) [reply shell/platform/darwin/ios/framework/Source/FlutterEngine.mm:310]: Totally fair :)

To save your brain work later, it's because of this:
> If your app targets iOS 9.0 and later or macOS 10.11 and later, and you used [addObserver:selector:name:object:](https://developer.apple.com/documentation/foundation/nsnotificationcenter/1415360-addobserver), you do not need to unregister the observer. If you forget or are unable to remove the observer, the system cleans up the next time it would have posted to it.

https://developer.apple.com/documentation/foundation/nsnotificationcenter/1413994-removeobserver

**Pull request merged**
